### ISSUE TITLE
MPI_Intercomm_merge never returns. Just hangs forever.

### ISSUE THREAD

**Issue by ricardopieper** [open-mpi/ompi#7901]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
Started to happen in all versions since 2.1.1, Tested 4.0.4, 4.0.2 and 2.1.1 (from sudo apt-get install libopenmpi2). It was working fine on all of them before.

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Tarball, followed basic instructions in the git readme. Downloaded 4.0.4 tarball from here: https://www.open-mpi.org/software/ompi/v4.0/ 

### Please describe the system on which you are running

* Operating system/version: Elementary OS Hera (Ubuntu 18.04)
* Computer hardware: Intel Core i5 6400, 16GB RAM
* Network type: Localhost

-----------------------------

## Details of the problem

I have this function that spawns a number of new processes and joins them all into a single communicator, one by one. It spawns a new process, merges it into a common communicator, and repeats. It works on a super old OpenMPI 1.4.5 on Ubuntu, OpenMPI 4.0.1 on my Macbook, and it used to work on OpenMPI 4.0.2 on my machine with Elementary OS until 3 or 4 weeks ago.

While the quality of this function is questionable at best (i don't know a better way to join a ton of processes into a single comm by calling spawn once, but I haven't revisited this problem since 6 months ago), it did the job for a long time, and still does in other computers. I know it looks kinda crazy, but I wish it still worked as before. I've tried reinstalling OpenMPI multiple times, different versions, and none of them work anymore.

It hangs on MPI_Intercomm_merge, which is a collective operation. Both the MPI_COMM_NULL and children hang when calling it for the first time. This is the code, i marked the hangs with a "This hangs!" comment. All new processes immediately call this function, and they all pass the same `numberOfProcesses` argument.
```
MPI_Comm SetTotalNumberOfProcesses(int argc, char **argv, int numberOfProcesses)
{
    
    MPI_Init(&argc, &argv);

    MPI_Comm parent;
    MPI_Comm_get_parent(&parent);

    if (parent == MPI_COMM_NULL)
    {
        MPI_Comm intracomm;

        std::cout << "Parent calling spawn" << std::endl;

        MPI_Comm process = spawn(argv, MPI_COMM_WORLD);

        std::cout << "Parent calling MPI_Intercomm_merge" << std::endl;

        //This hangs!
        MPI_Intercomm_merge(process, 0, &intracomm);

        for (int i = 0; i < numberOfProcesses - 1; i++)
        {
            int commSize;
            std::cout << "Parent calling MPI_Comm_size" << std::endl;

            MPI_Comm_size(intracomm, &commSize);

            for (int j = 1; j < commSize; j++)
            {
                int command = COMMAND_SPAWN_NEW;
                MPI_Send(&command, 1, MPI_INT, j, 0, intracomm);
            }
            std::cout << "Parent spawning.." << std::endl;
            MPI_Comm worker = spawn(argv, intracomm);
            MPI_Intercomm_merge(worker, 0, &intracomm);
        }

        {
            int commSize;
            MPI_Comm_size(intracomm, &commSize);
            for (int j = 1; j < commSize; j++)
            {
                int command = COMMAND_SPAWN_ENDED;
                std::cout << "Parent sending spawn ended" << std::endl;

                MPI_Send(&command, 1, MPI_INT, j, 0, intracomm);
            }
        }

        MPI_Barrier(intracomm);
        return intracomm;
    }
    else
    {
        MPI_Comm intracomm;

        std::cout << "Child calling MPI_Intercomm_merge!" << std::endl;
        
        //This hangs!
        MPI_Intercomm_merge(parent, 1, &intracomm);
        std::cout << "Child ended calling MPI_Intercomm_merge!" << std::endl;

        int myNewRank = -1;
        MPI_Comm_rank(intracomm, &myNewRank);

        int command = 0;
        MPI_Status status;

        while (true)
        {
            MPI_Recv(&command, 1, MPI_INT, 0, MPI_ANY_TAG, intracomm, &status);

            if (command == COMMAND_SPAWN_NEW)
            {
                MPI_Comm spawned = spawn(argv, intracomm);
                                        
                MPI_Intercomm_merge(spawned, 0, &intracomm);
                int myNewRank2 = -1;
                MPI_Comm_rank(intracomm, &myNewRank2);
            }
            else if (command == COMMAND_SPAWN_ENDED)
            {
                MPI_Comm_rank(intracomm, &myNewRank);
                break;
            }
        }

        MPI_Barrier(intracomm);
        numberOfProcessesAlreadySet = true;
        return intracomm;
    }
}
```

I reiterate that this only happens on my computer for some reason, on many versions of OpenMPI, and it started recently. Even on my Macbook this works. 

I'm asking for help to diagnose this issue. Where can I start? What should I look for?

**Comment by devreal** [comment]
@ricardopieper I tried to reproduce this but I fail to compile your example:

```
$ mpicxx mpi_intercomm_merge.cc -o mpi_intercomm_merge
mpi_intercomm_merge.cc: In function ‘ompi_communicator_t* SetTotalNumberOfProcesses(int, char**, int)’:
mpi_intercomm_merge.cc:18:28: error: ‘spawn’ was not declared in this scope
         MPI_Comm process = spawn(argv, MPI_COMM_WORLD);
                            ^~~~~
mpi_intercomm_merge.cc:34:31: error: ‘COMMAND_SPAWN_NEW’ was not declared in this scope
                 int command = COMMAND_SPAWN_NEW;
                               ^~~~~~~~~~~~~~~~~
mpi_intercomm_merge.cc:47:31: error: ‘COMMAND_SPAWN_ENDED’ was not declared in this scope
                 int command = COMMAND_SPAWN_ENDED;
                               ^~~~~~~~~~~~~~~~~~~
mpi_intercomm_merge.cc:77:28: error: ‘COMMAND_SPAWN_NEW’ was not declared in this scope
             if (command == COMMAND_SPAWN_NEW)
                            ^~~~~~~~~~~~~~~~~
mpi_intercomm_merge.cc:79:36: error: ‘spawn’ was not declared in this scope
                 MPI_Comm spawned = spawn(argv, intracomm);
                                    ^~~~~
mpi_intercomm_merge.cc:79:36: note: suggested alternative: ‘spawned’
                 MPI_Comm spawned = spawn(argv, intracomm);
                                    ^~~~~
                                    spawned
mpi_intercomm_merge.cc:85:33: error: ‘COMMAND_SPAWN_ENDED’ was not declared in this scope
             else if (command == COMMAND_SPAWN_ENDED)
                                 ^~~~~~~~~~~~~~~~~~~
mpi_intercomm_merge.cc:93:9: error: ‘numberOfProcessesAlreadySet’ was not declared in this scope
         numberOfProcessesAlreadySet = true;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
mpi_intercomm_merge.cc:93:9: note: suggested alternative: ‘numberOfProcesses’
         numberOfProcessesAlreadySet = true;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
         numberOfProcesses

```

Could you please post a minimal working example that compiles cleanly? That would make things easier for people to look into this.

**Comment by ricardopieper** [comment]
Hi @devreal sorry, my bad. This one should compile with mpic++.

It should also spit a ton of logs, but all it does in my machine nowadays is this:

```
Parent calling spawn
Parent calling MPI_Intercomm_merge
Child calling MPI_Intercomm_merge!
```

Here is the code:

```
#include "mpi.h"
#include <iostream>
#define COMMAND_SPAWN_NEW 1
#define COMMAND_SPAWN_ENDED 2

MPI_Comm spawn(char **argv, MPI_Comm comm)
{
    MPI_Comm intercomm;
    MPI_Comm_spawn(argv[0], &argv[1], 1, MPI_INFO_NULL, 0, comm, &intercomm, MPI_ERRCODES_IGNORE);
    return intercomm;
}

int main(int argc, char **argv)
{

    int numberOfProcesses = 5;
    MPI_Comm finalComm;

    MPI_Init(&argc, &argv);

    MPI_Comm parent;
    MPI_Comm_get_parent(&parent);

    if (parent == MPI_COMM_NULL)
    {
        MPI_Comm intracomm;

        std::cout << "Parent calling spawn" << std::endl;

        MPI_Comm process = spawn(argv, MPI_COMM_WORLD);

        std::cout << "Parent calling MPI_Intercomm_merge" << std::endl;

        //This hangs!
        MPI_Intercomm_merge(process, 0, &intracomm);

        for (int i = 0; i < numberOfProcesses - 1; i++)
        {
            int commSize;
            std::cout << "Parent calling MPI_Comm_size" << std::endl;

            MPI_Comm_size(intracomm, &commSize);

            for (int j = 1; j < commSize; j++)
            {
                int command = COMMAND_SPAWN_NEW;
                MPI_Send(&command, 1, MPI_INT, j, 0, intracomm);
            }
            std::cout << "Parent spawning.." << std::endl;
            MPI_Comm worker = spawn(argv, intracomm);
            MPI_Intercomm_merge(worker, 0, &intracomm);
        }

        {
            int commSize;
            MPI_Comm_size(intracomm, &commSize);
            for (int j = 1; j < commSize; j++)
            {
                int command = COMMAND_SPAWN_ENDED;
                std::cout << "Parent sending spawn ended" << std::endl;

                MPI_Send(&command, 1, MPI_INT, j, 0, intracomm);
            }
        }

        MPI_Barrier(intracomm);
        finalComm = intracomm;
    }
    else
    {
        MPI_Comm intracomm;

        std::cout << "Child calling MPI_Intercomm_merge!" << std::endl;
        
        //This hangs!
        MPI_Intercomm_merge(parent, 1, &intracomm);
        std::cout << "Child ended calling MPI_Intercomm_merge!" << std::endl;

        int myNewRank = -1;
        MPI_Comm_rank(intracomm, &myNewRank);

        int command = 0;
        MPI_Status status;

        while (true)
        {
            MPI_Recv(&command, 1, MPI_INT, 0, MPI_ANY_TAG, intracomm, &status);

            if (command == COMMAND_SPAWN_NEW)
            {
                MPI_Comm spawned = spawn(argv, intracomm);
                                        
                MPI_Intercomm_merge(spawned, 0, &intracomm);
                int myNewRank2 = -1;
                MPI_Comm_rank(intracomm, &myNewRank2);
            }
            else if (command == COMMAND_SPAWN_ENDED)
            {
                MPI_Comm_rank(intracomm, &myNewRank);
                break;
            }
        }

        MPI_Barrier(intracomm);
        finalComm = intracomm;
    }

    int myRank;
    MPI_Comm_rank(finalComm, &myRank);

    std::cout << "Reached end! Rank = " << myRank << std::endl;
    MPI_Finalize();
}
```

**Comment by devreal** [comment]
Thanks. I'm having trouble reproducing your hangs. I built Open MPI with UCX 1.8.0 enabled on local Linux Mint 19. I get the following output: 

```
$ mpirun -n 1 --oversubscribe ./mpi_intercomm_merge
Parent calling spawn
Parent calling MPI_Intercomm_merge
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Child ended calling MPI_Intercomm_merge!
Parent spawning..
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Parent spawning..
Child ended calling MPI_Intercomm_merge!
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Parent spawning..
Child ended calling MPI_Intercomm_merge!
Child calling MPI_Intercomm_merge!
Child ended calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Parent spawning..
Child calling MPI_Intercomm_merge!
Parent sending spawn ended
Parent sending spawn ended
Parent sending spawn ended
Parent sending spawn ended
Parent sending spawn ended
Child ended calling MPI_Intercomm_merge!
Reached end! Rank = 0
Reached end! Rank = 4
Reached end! Rank = 1
Reached end! Rank = 2
Reached end! Rank = 3
Reached end! Rank = 5
```

However, during `MPI_Finalize` I get the an error from UCX error and and things go up in smoke:
```
[beryl:14085:0:14085]      ucp_ep.c:725  Bug: pending request 0x55f392668ce8 on ep 0x7f97a008c048 should have been flushed

/home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_ep.c: [ ucp_destroyed_ep_pending_purge() ]
      ...
      722 }
      723 
      724 static unsigned ucp_ep_do_disconnect(void *arg)
==>   725 {
      726     ucp_request_t *req = arg;
      727 
      728     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_COMPLETED));

==== backtrace (tid:  14386) ====
 0 0x0000000000026375 ucs_debug_print_backtrace()  /home/joseph/src/ucx/ucx/build/src/ucs/../../../src/ucs/debug/debug.c:625
 1 0x000000000001953b ucp_destroyed_ep_pending_purge()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_ep.c:725
 2 0x000000000005bc36 ucp_wireup_ep_pending_purge()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/wireup/wireup_ep.c:227
 3 0x000000000005bc36 ucp_wireup_ep_pending_purge()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/wireup/wireup_ep.c:222
 4 0x000000000001961d uct_ep_pending_purge()  /home/joseph/src/ucx/ucx/build/../src/uct/api/uct.h:2607
 5 0x000000000001961d ucp_ep_cleanup_lanes()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_ep.c:746
 6 0x0000000000019558 ucp_ep_destroy_internal()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_ep.c:731
 7 0x0000000000028421 ucp_worker_destroy_eps()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_worker.c:1855
 8 0x0000000000028421 ucp_worker_destroy()  /home/joseph/src/ucx/ucx/build/src/ucp/../../../src/ucp/core/ucp_worker.c:1876
 9 0x000000000000636b mca_pml_ucx_cleanup()  /home/joseph/src/openmpi/openmpi-4.0.4/build/ompi/mca/pml/ucx/../../../../../ompi/mca/pml/ucx/pml_ucx.c:354
10 0x000000000004fe1c ompi_mpi_finalize()  /home/joseph/src/openmpi/openmpi-4.0.4/build/ompi/../../ompi/runtime/ompi_mpi_finalize.c:342
11 0x00000000000012ef main()  ???:0
12 0x0000000000021b97 __libc_start_main()  /build/glibc-2ORdQG/glibc-2.27/csu/../csu/libc-start.c:310
13 0x0000000000000dca _start()  ???:0
=================================
```

Might this be a bug in the UCX PML @jladd-mlnx?

With the UCX PML disabled, things look better (still seems to be an issue with a broken pipe):
```
$ mpirun -n 1 --oversubscribe --mca pml ^ucx ./mpi_intercomm_merge
Parent calling spawn
Parent calling MPI_Intercomm_merge
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Child ended calling MPI_Intercomm_merge!
Parent spawning..
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Child ended calling MPI_Intercomm_merge!
Parent spawning..
Child calling MPI_Intercomm_merge!
Child ended calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Parent spawning..
Child calling MPI_Intercomm_merge!
Parent calling MPI_Comm_size
Parent spawning..
Child ended calling MPI_Intercomm_merge!
Child calling MPI_Intercomm_merge!
Parent sending spawn ended
Child ended calling MPI_Intercomm_merge!
Parent sending spawn ended
Parent sending spawn ended
Parent sending spawn ended
Parent sending spawn ended
Reached end! Rank = 1
Reached end! Rank = 3
Reached end! Rank = 5
Reached end! Rank = 0
Reached end! Rank = 2
Reached end! Rank = 4
--------------------------------------------------------------------------
Child job 2 terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node beryl exited on signal 13 (Broken pipe).
--------------------------------------------------------------------------
```

(Note: I added a missing `return 0;` after `MPI_Finalize` in the reproducer code)

However, in both cases at least I get past the merge. Could you please post your configure line for Open MPI 4.0.4? And can you try to attach to your process(es) to get a backtrace from gdb to see where they are stuck?

**Comment by devreal** [comment]
Tried with current `master`, both issues are gone there.

**Comment by ricardopieper** [comment]
@devreal hmm, my configure line? Can you clarify?

If you mean the `./configure --prefix=<path>`, I don't remember what I did in this step. I just did the most basic thing possible probably, i.e. just `./configure --prefix=/home/ricardo/openmpi4`. However, this happens on the openmpi 2 version as well, the one you can install via `apt-get install` on Ubuntu 18/Elementary OS Hera.

If you mean the command I use to run the program, I compile it with  `mpic++ mpibug.cpp -o mpibug` and then `mpirun -np 1 --oversubscribe mpibug`.

I also used to get this `signal 13 (Broken pipe)` line all the time on my PC. I have no such issue on the clusters I actually run this code though.

**Comment by devreal** [comment]
I installed Open MPI 2.1.1 (which was released over 3 years ago [1]) from apt and I get the same result as with 4.0.4 (and current 4.0.x branch for that matter) without the UCX PML: it runs to completion but exits with Signal 13. Not sure what could be the difference between my system and yours.

[1] https://www.open-mpi.org/software/ompi/versions/timeline.php

**Comment by ricardopieper** [comment]
Yeah, if only I knew the reason why it hangs I could do something about it. But I'm not exactly blocked by this issue.

I think some system-wide change (like an apt-get upgrade) caused it. I had the exact same results as you a few weeks ago.

My goal with that code is to support autonomic computing, where resources are added or removed as needed... but to be honest, this isn't exactly urgent or important for this stage of the research. If no one else is having this issue then I think it can be ignored, though it is kinda weird that it just stopped working *for all versions* of MPI.

For now I'm just enabled static process allocation with `-np "number of processes"` like a normal human being would do. I just figured I could let people know that this happens.

**Comment by ricardopieper** [comment]
Btw, this is the stack trace of the initially spawned process via mpirun, on OpenMPI 4.0.4:

```
#0  0x00007f49c7410cf9 in __GI___poll (fds=fds@entry=0x5577c82cdb00, nfds=nfds@entry=4, timeout=timeout@entry=0) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007f49c6dba3e3 in poll (__timeout=<optimized out>, __nfds=4, __fds=0x5577c82cdb00) at /usr/include/x86_64-linux-gnu/bits/poll2.h:46
#2  poll_dispatch (base=0x5577c81f93f0, tv=0x7fff599ac1f0) at poll.c:165
#3  0x00007f49c6db15fb in opal_libevent2022_event_base_loop (base=0x5577c81f93f0, flags=2) at event.c:1630
#4  0x00007f49c6d62875 in opal_progress () from /home/ricardo/openmpi4/lib/libopen-pal.so.40
#5  0x00007f49c7aaa085 in ompi_comm_nextcid () from /home/ricardo/openmpi4/lib/libmpi.so.40
#6  0x00007f49c7af50fe in PMPI_Intercomm_merge () from /home/ricardo/openmpi4/lib/libmpi.so.40
#7  0x00005577c7d7cfe2 in main (argc=1, argv=0x7fff599ac548) at mpibug.cpp:36
```

This is the stack trace of the process spawned via MPI_Spawn, basically the same thing:

```
#0  0x00007fae8baa9cf9 in __GI___poll (fds=fds@entry=0x562266379020, nfds=nfds@entry=4, timeout=timeout@entry=0) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007fae8b4533e3 in poll (__timeout=<optimized out>, __nfds=4, __fds=0x562266379020) at /usr/include/x86_64-linux-gnu/bits/poll2.h:46
#2  poll_dispatch (base=0x5622662a3d50, tv=0x7ffc6ad37420) at poll.c:165
#3  0x00007fae8b44a5fb in opal_libevent2022_event_base_loop (base=0x5622662a3d50, flags=2) at event.c:1630
#4  0x00007fae8b3fb875 in opal_progress () from /home/ricardo/openmpi4/lib/libopen-pal.so.40
#5  0x00007fae8c143085 in ompi_comm_nextcid () from /home/ricardo/openmpi4/lib/libmpi.so.40
#6  0x00007fae8c18e0fe in PMPI_Intercomm_merge () from /home/ricardo/openmpi4/lib/libmpi.so.40
#7  0x00005622658b91ad in main (argc=1, argv=0x7ffc6ad37778) at mpibug.cpp:77
```

Meanwhile, 2 of my cores are pinned at 100%, though it shows 50% of the time is spent in the kernel.. as evidenced by the stacktrace itself it seems.

Maybe it's an issue on libevent?

**Comment by rhc54** [comment]
FWIW: we do see that signature in the community IBM nightly regression tests - but on no other platform. It is an issue with the `ompi_comm_nextcid` routine - nobody over there has yet had time to investigate.

**Comment by ricardopieper** [comment]
Also, on the MPI_Spawn'ed process, if I enter `next` on LLDB (not gdb) a bunch of times (sorry i'm a real noob at LLDB) this is what happens. Maybe this can be useful somehow:

```
(lldb) process attach --pid 11841
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x000055e764a9f060, nfds=4, timeout=0) at poll.c:29
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30

Executable module set to "/home/ricardo/mpibug".
Architecture set to: x86_64-pc-linux.
(lldb) thread backtrace
* thread #1, name = 'mpibug', stop reason = signal SIGSTOP
  * frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x000055e764a9f060, nfds=4, timeout=0) at poll.c:29
    frame #1: 0x00007f9188ea33e3 libopen-pal.so.40`poll_dispatch [inlined] poll(__timeout=<unavailable>, __nfds=4, __fds=0x000055e764a9f060) at poll2.h:46
    frame #2: 0x00007f9188ea33d6 libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:165
    frame #3: 0x00007f9188e9a5fb libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1630
    frame #4: 0x00007f9188e4b875 libopen-pal.so.40`opal_progress + 245
    frame #5: 0x00007f9189b93085 libmpi.so.40`ompi_comm_nextcid + 261
    frame #6: 0x00007f9189bde0fe libmpi.so.40`MPI_Intercomm_merge + 942
    frame #7: 0x000055e7645241ad mpibug`main(argc=1, argv=0x00007ffd685cd9e8) at mpibug.cpp:77
    frame #8: 0x00007f9189406b97 libc.so.6`__libc_start_main(main=(mpibug`main at mpibug.cpp:15), argc=1, argv=0x00007ffd685cd9e8, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007ffd685cd9d8) at libc-start.c:310
    frame #9: 0x000055e764523dba mpibug`_start + 42
(lldb) step
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step in
    frame #0: 0x00007f91895158f6 libc.so.6`__libc_disable_asynccancel at cancellation.S:89
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb) next
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f91895158f8 libc.so.6`__libc_disable_asynccancel at cancellation.S:91
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515900 libc.so.6`__libc_disable_asynccancel at cancellation.S:92
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515903 libc.so.6`__libc_disable_asynccancel at cancellation.S:93
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515907 libc.so.6`__libc_disable_asynccancel at cancellation.S:94
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515912 libc.so.6`__libc_disable_asynccancel at cancellation.S:96
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515914 libc.so.6`__libc_disable_asynccancel at cancellation.S:98
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9189515917 libc.so.6`__libc_disable_asynccancel at cancellation.S:99
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f918951591a libc.so.6`__libc_disable_asynccancel at cancellation.S:100
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f918951591d libc.so.6`__libc_disable_asynccancel at cancellation.S:101
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f918951591f libc.so.6`__libc_disable_asynccancel at cancellation.S:102
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f91894f9d0d libc.so.6`__GI___poll(fds=0x000055e764a9f060, nfds=4, timeout=0) at poll.c:29
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f91894f9d11 libc.so.6`__GI___poll(fds=0x000055e764a9f060, nfds=4, timeout=0) at poll.c:43
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea33e3 libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:167
   164 	
   165 		res = poll(event_set, nfds, msec);
   166 	
-> 167 		EVBASE_ACQUIRE_LOCK(base, th_base_lock);
   168 	
   169 		if (res == -1) {
   170 			if (errno != EINTR) {
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea33eb libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:165
   162 	
   163 		EVBASE_RELEASE_LOCK(base, th_base_lock);
   164 	
-> 165 		res = poll(event_set, nfds, msec);
   166 	
   167 		EVBASE_ACQUIRE_LOCK(base, th_base_lock);
   168 	
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea33ed libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:167
   164 	
   165 		res = poll(event_set, nfds, msec);
   166 	
-> 167 		EVBASE_ACQUIRE_LOCK(base, th_base_lock);
   168 	
   169 		if (res == -1) {
   170 			if (errno != EINTR) {
  thread #2, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f91894f9cf9 libc.so.6`__GI___poll(fds=0x00007f917c000b20, nfds=1, timeout=3599986) at poll.c:29
  thread #3, name = 'mpibug', stop reason = signal SIGSTOP
    frame #0: 0x00007f9189506d67 libc.so.6`epoll_wait(epfd=10, events=0x000055e764a3ba80, maxevents=32, timeout=2100000) at epoll_wait.c:30
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea33fe libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:169
   166 	
   167 		EVBASE_ACQUIRE_LOCK(base, th_base_lock);
   168 	
-> 169 		if (res == -1) {
   170 			if (errno != EINTR) {
   171 				event_warn("poll");
   172 				return (-1);
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea3407 libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:180
   177 	
   178 		event_debug(("%s: poll reports %d", __func__, res));
   179 	
-> 180 		if (res == 0 || nfds == 0)
   181 			return (0);
   182 	
   183 		i = random() % nfds;
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea34ca libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:175
   172 				return (-1);
   173 			}
   174 	
-> 175 			return (0);
   176 		}
   177 	
   178 		event_debug(("%s: poll reports %d", __func__, res));
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188ea34cc libopen-pal.so.40`poll_dispatch(base=0x000055e7649c9d50, tv=0x00007ffd685cd690) at poll.c:208
   205 		}
   206 	
   207 		return (0);
-> 208 	}
   209 	
   210 	static int
   211 	poll_add(struct event_base *base, int fd, short old, short events, void *_idx)
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a5fb libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1632
   1629	
   1630			res = evsel->dispatch(base, tv_p);
   1631	
-> 1632			if (res == -1) {
   1633				event_debug(("%s: dispatch returned unsuccessfully.",
   1634					__func__));
   1635				retval = -1;
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a604 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1639
   1636				goto done;
   1637			}
   1638	
-> 1639			update_time_cache(base);
   1640	
   1641			timeout_process(base);
   1642	
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a61d libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1641
   1638	
   1639			update_time_cache(base);
   1640	
-> 1641			timeout_process(base);
   1642	
   1643			if (N_ACTIVE_CALLBACKS(base)) {
   1644				int n = event_process_active(base);
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a62d libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1643
   1640	
   1641			timeout_process(base);
   1642	
-> 1643			if (N_ACTIVE_CALLBACKS(base)) {
   1644				int n = event_process_active(base);
   1645				if ((flags & EVLOOP_ONCE)
   1646				    && N_ACTIVE_CALLBACKS(base) == 0
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a641 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1649
   1646				    && N_ACTIVE_CALLBACKS(base) == 0
   1647				    && n != 0)
   1648					done = 1;
-> 1649			} else if (flags & EVLOOP_NONBLOCK)
   1650				done = 1;
   1651		}
   1652		event_debug(("%s: asked to terminate loop.", __func__));
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a665 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1562
   1559		const struct eventop *evsel = base->evsel;
   1560		struct timeval tv;
   1561		struct timeval *tv_p;
-> 1562		int res, done, retval = 0;
   1563	
   1564		/* Grab the lock.  We will release it inside evsel.dispatch, and again
   1565		 * as we invoke user callbacks. */
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a667 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1658
   1655		clear_time_cache(base);
   1656		base->running_loop = 0;
   1657	
-> 1658		EVBASE_RELEASE_LOCK(base, th_base_lock);
   1659	
   1660		return (retval);
   1661	}
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a66e libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1655
   1652		event_debug(("%s: asked to terminate loop.", __func__));
   1653	
   1654	done:
-> 1655		clear_time_cache(base);
   1656		base->running_loop = 0;
   1657	
   1658		EVBASE_RELEASE_LOCK(base, th_base_lock);
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a679 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1656
   1653	
   1654	done:
   1655		clear_time_cache(base);
-> 1656		base->running_loop = 0;
   1657	
   1658		EVBASE_RELEASE_LOCK(base, th_base_lock);
   1659	
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a684 libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1658
   1655		clear_time_cache(base);
   1656		base->running_loop = 0;
   1657	
-> 1658		EVBASE_RELEASE_LOCK(base, th_base_lock);
   1659	
   1660		return (retval);
   1661	}
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e9a49c libopen-pal.so.40`opal_libevent2022_event_base_loop(base=0x000055e7649c9d50, flags=2) at event.c:1661
   1658		EVBASE_RELEASE_LOCK(base, th_base_lock);
   1659	
   1660		return (retval);
-> 1661	}
   1662	
   1663	/* Sets up an event for processing once */
   1664	struct event_once {
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = step over
    frame #0: 0x00007f9188e4b875 libopen-pal.so.40`opal_progress + 245
libopen-pal.so.40`opal_progress:
->  0x7f9188e4b875 <+245>: movl   $0x0, 0x2db30d(%rip)      ; opal_progress_yield_when_idle
    0x7f9188e4b87f <+255>: movq   0x2d4a32(%rip), %rax
    0x7f9188e4b886 <+262>: cmpb   $0x0, (%rax)
    0x7f9188e4b889 <+265>: je     0x7f9188e4b88f            ; <+271>
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = instruction step over
    frame #0: 0x00007f9188e4b87f libopen-pal.so.40`opal_progress + 255
libopen-pal.so.40`opal_progress:
->  0x7f9188e4b87f <+255>: movq   0x2d4a32(%rip), %rax
    0x7f9188e4b886 <+262>: cmpb   $0x0, (%rax)
    0x7f9188e4b889 <+265>: je     0x7f9188e4b88f            ; <+271>
    0x7f9188e4b88b <+267>: testl  %ebp, %ebp
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = instruction step over
    frame #0: 0x00007f9188e4b886 libopen-pal.so.40`opal_progress + 262
libopen-pal.so.40`opal_progress:
->  0x7f9188e4b886 <+262>: cmpb   $0x0, (%rax)
    0x7f9188e4b889 <+265>: je     0x7f9188e4b88f            ; <+271>
    0x7f9188e4b88b <+267>: testl  %ebp, %ebp
    0x7f9188e4b88d <+269>: jle    0x7f9188e4b8c0            ; <+320>
(lldb)  
Process 11841 stopped
* thread #1, name = 'mpibug', stop reason = instruction step over
    frame #0: 0x00007f9188e4b889 libopen-pal.so.40`opal_progress + 265
libopen-pal.so.40`opal_progress:
->  0x7f9188e4b889 <+265>: je     0x7f9188e4b88f            ; <+271>
    0x7f9188e4b88b <+267>: testl  %ebp, %ebp
    0x7f9188e4b88d <+269>: jle    0x7f9188e4b8c0            ; <+320>
    0x7f9188e4b88f <+271>: addq   $0x8, %rsp
(lldb)  
(lldb) 
```

**Comment by rhc54** [comment]
Not really useful, I'm afraid - we know it is sitting in the event loop waiting for data to be returned. What we don't know is why that data isn't showing up.

**Comment by ricardopieper** [comment]
Ah, shame...

If I find anything else, I'll let you know. 
Also if you guys want me to try something out, just let me know as well.

**Comment by bosilca** [comment]
Quick note here: the SM or vader BTL do not support communications between processes between different job id, even when they are on the same node. So, I think you need to explicitly enable tcp for local communications (which is disabled by default).

**Comment by ricardopieper** [comment]
@bosilca I tried this command:
`mpirun --mca btl tcp,self --mca pml ob1 -n 1 --oversubscribe mpibug`

But the same thing happens. It used to work without the ` --mca btl tcp,self --mca pml ob1` part.

**Comment by bosilca** [comment]
@ricardopieper I got a mixed bag of outcomes. I tried 4.0, 4.1 and master, under the assumption that the host where the child processes will be spawned is important. I therefore increased the number of call to spawn to a number larger than the number of cores on the first host of the allocation (oversubscribe param is then necessary).

In 4.0 and 4.1 have a similar outcome, the app works as long as the spawn happen on the same node. As soon as you cross the node boundary, neither UCX or tcp works. The error is somewhere in ompi_dpm_dyn_init, where apparently OMPI fails to correctly receive the info from the child on a different node.

The good news is that master seems to behave in a much better way. I was able to heavily overload my allocation both with UCX and with a carefully configured TCP.
- with UCX I get the same issue as @devreal, pending requests are still on some internal UCX queues during MPI_Finalize (and those requests do not seem to be generated by the MPI app).
- with TCP, the problem is that on machines with multiple networks we are still not correctly detecting a valid peer, and instead drop legitimate connections attempts. Most of the cases triggered the error handler, but I did noticed few where it seems that the application deadlocked. If I limit the number of interface to a single one per node, I can run your application successfully with TCP.

**Comment by jsquyres** [comment]
Potentially also related to/a duplicate of #7620.

**Comment by rhc54** [comment]
Extensive investigation on OMPI master cannot replicate this problem - we have concluded that recent changes in PRRTE have fixed it there. This has been included in the upcoming v5 release, which should be coming along soon. In the interim, you are welcome to try one of the nightly tarballs from either the master or v5 branch (https://www.open-mpi.org/nightly/). Please reopen if you continue to see this problem there.
